Merged depending fixes for IZPACK-1260 and  IZPACK-1406

rkrell opened the pull request:
[IZPACK-1260](https://izpack.atlassian.net/browse/IZPACK-1260): Dynamic variables refresh with conditions depending on the dynamic variables does not use current variable values:
- fixed in DefaultVariables.refresh()
- added testcase for this
  [IZPACK-1406](https://izpack.atlassian.net/browse/IZPACK-1406): Preorder dynamic variables according to references among them at compile time:
- cherry-picked/merged and adapted 8bb7f56, 18f21b4, 74ab1c8 and 957d9d4 from [branch Helpstone/izpack/tree/IZPACK-1406_with_conditions](https://github.com/Helpstone/izpack/tree/IZPACK-1406_with_conditions)
- Enhanced console installer testcase by standalone check-once dynamic variable without unresolved references (has not been covered in any test)
- Improved recognition of unresolved variables during variables refresh (regular expression)

rkrell committed:
IZPACK-1260: Dynamic variables refresh with conditions depending on the dynamic variables does not use current variable values:
 - fixed in DefaultVariables.refresh()
 - added testcase for this
IZPACK-1406: Preorder dynamic variables according to references among them at compile time:
- cherry-picked and adapted 8bb7f56, 18f21b4, 74ab1c8 and 957d9d4 from Helpstone/izpack/tree/IZPACK-1406_with_conditions
- Enhanced console installer testcase by standalone check-once dynamic variable without unresolved references (has not been covered in any test)
- Improved recognition of unresolved variables during variables refresh (regex)

rkrell committed:
Merge remote-tracking branch 'helpstone/IZPACK-1406_with_conditions' into IZPACK-1406

Helpstone commented on izpack-core/src/main/java/com/izforge/izpack/core/data/DefaultVariables.java:
```diff
@@ -294,89 +296,133 @@ public synchronized void add(DynamicVariable variable)
+        int maxCount = 10*dynamicVariables.size()+1;
+        int count=maxCount;
+        boolean changed = true;
+        while (changed) {
```
The intention of this branch IZPACK-1406_with_conditions instead of the previous IZPACK-1406 was to get a deterministic single loop refresh() by taking the variables used in conditions into account.

Now you've switched back to the old behaviour here. Probably this is an unintented change? Isnt't it?

rkrell replied:
> The intention of this branch IZPACK-1406_with_conditions instead of the previous IZPACK-1406 was to get a deterministic single loop refresh() by taking the variables used in conditions into account.
> Now you've switched back to the old behaviour here. Probably this is an unintented change? Isnt't it?

Yes and no. I've tested it before in this way, but I'll do some tests without the loop.
Anyway, if the elements will be well-ordered, the only disadvantage left with the loop should be some more lines of code and one comparison. But it will be better readable without it, of course.

If it will work without the loop, I'll update the request. Just give me some time to check this.

rkrell replied:
I'm still doing tests, but not to block you from further changes - I removed the loop in commit f6b1a5688c5de5638064fde06b9d74317ba811e2.

rkrell committed:
IZPACK-1406: Removed cyclic variable refresh, dynamic variables are now completely pre-ordered

rkrell commented:
I tested it "as is" with the removed refresh cycles also in complex installers.
Your implementation of pre-ordering seems to work very well.
From my point of view this is ready to be merged.

I'd still discuss my complaints regards the changed Condition interface, posted directly after the dedicated lines of source code in the commit diff. Thanks.

Helpstone commented on izpack-core/src/test/java/com/izforge/izpack/core/data/DefaultVariablesTest.java:
```diff
@@ -240,10 +246,56 @@ public void testConditionalDynamicVariables()
+
+        variables.refresh();
+
+        assertEquals(confFile.getAbsolutePath(), variables.get("previous.wrapper.conf"));
```
On windows this test does fail:

> org.junit.ComparisonFailure: expected:<...59586752314062\myapp[\conf]wrapper.conf> but was:<...59586752314062\myapp[/conf/]wrapper.conf>

The file.separator is wrong in this comparison!

rkrell replied:
I see. I'll fix that, ok?

Helpstone replied:
Yes, this would be great!
 
-----Ursprüngliche Nachricht-----
 Von: "René Krell" [notifications@github.com]
 Gesendet: Di. 21.06.2016 11:46
 An: "Helpstone/izpack" [izpack@noreply.github.com]
 Kopie: "Helpstone" [dth@freenet.de],  "Comment" [comment@noreply.github.com]
 Betreff: Re: [Helpstone/izpack] Merged depending fixes for IZPACK-1260 and IZPACK-1406 (#1)

In izpack-core/src/test/java/com/izforge/izpack/core/data/DefaultVariablesTest.java:

> -            installPath = rootFoldernewFolder("myapp");> +            File confPath = new File(installPath, "conf");> +            confPath.mkdirs();> +            confFile = new File(confPath, "wrapper.conf");> +            FileUtils.touch(confFile);> +        }> +        catch (IOException e)> +        {> +            fail("File operation failed.");> +        }> +> +        variables.set("INSTALL_PATH", installPath.getAbsolutePath());> +> +        variables.refresh();> +> +        assertEquals(confFile.getAbsolutePath(), variables.get("previous.wrapper.conf"));I see. I'll fix that, ok?
>   —
>   You are receiving this because you commented.
>   Reply to this email directly, view it on GitHub, or mute the thread.
>   -----Ursprüngliche Nachricht Ende-----

---

Alle Postfächer an einem Ort. Jetzt wechseln und E-Mail-Adresse mitnehmen! Rundum glücklich mit freenetMail

rkrell replied:
> Yes, this would be great!

Should be fixed (in theory), see above, refresh the code from this PR and let me know whether it works for you.

Helpstone committed:
Make the name of installer.xml accessible in unit tests

Helpstone committed:
IZPACK-1406: implement some tests for correct ordering of dynamic var
computation

rkrell committed:
IZPACK-1260: Post-fix: Fixed unit test on Windows (wrong path separator in comparison)

Helpstone committed:
Fix: Do not use regex-replace, because it will have problems with
backslash

Helpstone committed:
IZPACK-1406: Do implement some test cases for complex variable values

Helpstone committed:
IZPACK-1406: Do parseUnresolvedVariableNames() only on the additional
vars, because otherwise found recursive dependencies
super.getUnresolvedVariableNames() are lost

Helpstone committed:
make code more readable

rkrell committed:
Merge pull request #5 from Helpstone/IZPACK-1406-RKrell

Fix: Do not use regex-replace, because it will have problems with backslash

rkrell committed:
Merge pull request #6 from Helpstone/IZPACK-1406

Izpack 1406 - some unit tests for variable order

rkrell committed:
Merge pull request #7 from Helpstone/IZPACK-1406d

Code Cleanup fopr IZPACK-1406
